core/src/services/aliyun_drive/writer.rs

```diff
@@ -34,7 +34,7 @@ pub struct AliyunDriveWriter {
     parent_file_id: String,
     name: String,
 
-    file_id: Arc<RwLock<Option<String>>>,
+    ids: Arc<RwLock<Option<(String, String)>>>,
```

Comment:
Hi, I find it a bit odd that an `Arc<RwLock<Option<(String, String)>>>` is stored in `AliyunDriveWriter`. Ideally, it should be stateless when calling `MultipartWrite`, as all states are managed by `MultipartWriter`.

Could you explain this part a bit?

Reply:
The state is used to store the file_id and upload_id. The Aliyun Drive API requires a file_id each time we fetch the upload_url for each part and complete the part, but I couldn't find it from the multipart writer, so I stored it.

I hadn't stored the upload_id before this PR. However, in #4780 I found that the retry behavior will cause some issues. Normally, we get the upload_id from `initiate_part` in the first call. When concurrency occurs, Aliyun Drive randomly fails on the method `initate_part` and has retries. Meanwhile, Aliyun Drive API doesn't return the upload_id after the first call on the same path. So I have to store it in case we lose it and finally encounter an error when encountering 'file exists'.

Reply:
IMO, retry behavior can be burdensome for methods consisting of multiple API calls, some of which can only be called once and require storing their result status.

Reply:
> Normally, we get the upload_id from `initiate_part` in the first call. When concurrency occurs, Aliyun Drive randomly fails on the method `write_once` which invokes `initiate_part` and has retries.

Oh, I see. The real issue here is that Aliyun Drive should not be treated as a valid multipart upload service.

A valid multipart upload service should be exactly the same with s3 that meets:

- An API such as `PutObject` can atomically create and upload a file.
- And a seperate group of APIs that can upload files in multiple parts: init, upload and complete.

Those two set of APIs should never be called in mixed. For example, `wirte_once` should never call `init_multipart`, it must be a single API call that create the file without extra states.

So back to our current issue. It's wrong to implement `MultipartWrite::write_once` by:

```rust
async fn write_once(&self, size: u64, body: crate::Buffer) -> Result<()> {
    let upload_id = self.initiate_part().await?;
    self.write_part(&upload_id, 0, size, body).await?;

    self.complete(&upload_id).await?;
    Ok(())
}
```

I suggest to implement `oio::Write` directly in such case.

Reply:
But Aliyun Drive requires to split a file into parts with maxium size of 5Gib.

Reply:
> But Aliyun Drive requires to split a file into parts with maxium size of 5Gib.

Implement `oio::Write` doesn't mean you can't use multipart upload. You can implement all current logic in `oio::Write` directly without lock:

```rust
pub struct AliyunDriveWriter {
    core: Arc<AliyunDriveCore>,

    _op: OpWrite,
    parent_file_id: String,
    name: String,

    file_id: Option<String>,
    upload_id: Option<String>,
    parts: Vec<AliyunDrivePart>
}
```

OpenDAL makes sure that all writing data are in order.

Reply:
Alright. Let me take a try.